My store page object

evgeniyamelone opened the pull request:
Implement an automated test using TestNG + Selenium for new account registration on http://automationpractice.com
Submit a link to the pull request (dev branch to master) with the completed test to this assignment.

Test scenario:
 
1.  Navigate to the target automationpractice.com; 
2. Open registration form; 
3. Fill all fields in the form with the valid data; 
4. Click 'Register'; 
5. Verify that new account is successfully created. 

Additional requirements:
 
* Target browser: Chrome 
* The test should be stable and may be executed multiple times without any false failures.


- page classes containing functionality for interacting with the page and verify data on it (use PageFactory pattern)
- class for storing account data 
- logging (slf4j)
- configuration property file to store the parameters required for the test 

Implement following improvements for your test framework:

- install Lombok (plugin and dependency) and apply Lombok annotations to your model class; 
- store test data in JSON data files; 
- simple data pool class to handle data files; 
- implement login\logout parameters for the test scenario 

Tests Implementation 
Implement tests for:
 - verification of account data for created account ( personal info and addresses ); 
- edit personal info and addresses.

evgeniyamelone committed:
Add:

- test for http://automationpractice.com/

evgeniyamelone committed:
Add:

- test for http://automationpractice.com/ with page objects

evgeniyamelone committed:
Add:

- test for http://automationpractice.com/ with page objects

evgeniyamelone committed:
Add:

- test for http://automationpractice.com/ with page objects
- lombok plugin

evgeniyamelone committed:
Add:

- test for http://automationpractice.com/ with page objects
- lombok plugin

evgeniyamelone committed:
Add:

- test for http://automationpractice.com/ with page objects
- lombok plugin

evgeniyamelone committed:
Remove extra file

mih-kopylov commented on pom.xml:
```diff
@@ -52,5 +61,25 @@
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.1</version>
```
best practice is to store versions in `<properties></properties` :)

evgeniyamelone replied:
I missed it =/

fixed

mih-kopylov commented on src/main/java/MyStoreRegistration.java:
```diff
@@ -0,0 +1,50 @@
+      //  driver.findElement(By.xpath("//*[@id=\"header\"]/div[2]/div/div/nav/div[1]/a")).click();
+        driver.findElement(By.className("login")).click();
+        WebDriverWait wait = new WebDriverWait(driver, 5);
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"email_create\"]")));
```
I believe, there's `By.id` locator as well? ;)

evgeniyamelone replied:
Oh, I forgot

evgeniyamelone replied:
Removed this file)

mih-kopylov commented on src/test/java/pageobjects/MainPage.java:
```diff
@@ -0,0 +1,24 @@
+
+public class MainPage {
+    private WebDriver driver;
+    private WebDriverWait wait;
```
I'm not sure that is used?

evgeniyamelone replied:
I forgot to remove it. Fixed

mih-kopylov commented on src/test/java/pageobjects/AccountCreationPage.java:
```diff
@@ -0,0 +1,87 @@
+    public AccountCreationPage(WebDriver driver) {
+        PageFactory.initElements(driver, this);
+        this.driver = driver;
+        this.wait = new WebDriverWait(driver, 30);
```
This seems not to be used as well?

evgeniyamelone replied:
I forgot to remove it. Fixed

mih-kopylov commented on src/test/java/ui/tests/RegistrationTest.java:
```diff
@@ -0,0 +1,82 @@
+    public static void setup() {
+        logger.info("Setup process started");
+        System.setProperty("webdriver.chrome.driver", ConfigProperties.getProperty("chromedriver"));
+        logger.info("ChromeDriver initiated");
```
it just gets initiated next string :)

evgeniyamelone replied:
sorry, I didn't get

mih-kopylov replied:
line 35: write to log: "driver initiated"
line 36: actual line when the driver gets initiated :D 

If the initialization fails, log will look a bit odd:
```
- driver initiated
- oops, it's actually failed
```

evgeniyamelone replied:
Updated

mih-kopylov commented on src/test/java/ui/tests/RegistrationTest.java:
```diff
@@ -0,0 +1,82 @@
+        accountPage = new AccountPage(driver);
+        logger.info("Setup implicitly wait for each method");
+        driver.manage().timeouts().implicitlyWait(15, TimeUnit.SECONDS);
+        driver.get(ConfigProperties.getProperty("mainpage"));
```
It's probably better to keep general configuration in `@BeforeClass` but move pages manipulation to either `@Before` method or in test, so that it's clear what page is used in each test. Even though there's a single test for now :)

evgeniyamelone replied:
Moved

evgeniyamelone committed:
Update:

- move versions to property

evgeniyamelone committed:
Delete:

- Unnecessary test

evgeniyamelone committed:
Update:

- moved page initiation to @BeforeEach

evgeniyamelone committed:
Update:

- moved page initiation to @Before

evgeniyamelone committed:
Remove:

- extra wait

evgeniyamelone committed:
Add:

- Declare Jackson

evgeniyamelone committed:
Add:

- Data Pool

evgeniyamelone committed:
Add:

- Data Pool usage in test

evgeniyamelone committed:
Update:

- Updated logging

evgeniyamelone committed:
Update:

- Test to update account is added

evgeniyamelone committed:
Update:

- Test to update account is added

mih-kopylov commented on pom.xml:
```diff
@@ -52,5 +64,30 @@
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>${lang3.version}</version>
```
should it be `commons-lang3.version`?

evgeniyamelone replied:
Updated

mih-kopylov commented on src/test/java/pageobjects/AccountCreationPage.java:
```diff
@@ -0,0 +1,86 @@
+        aliasField.sendKeys(userForm.getAlias());
+        submitAccountButton.click();
+
+        return new UserAccount(userForm.getFirstName(), userForm.getLastName(), userForm.getPassword(),
```
Didn't get the point - why creating a new instance of UserAccuont from another UserAccount?

ssheludchenko replied:
Next level UI verifications 👍

mih-kopylov commented on src/test/java/pageobjects/MainPage.java:
```diff
@@ -0,0 +1,22 @@
+public class MainPage {
+    private WebDriver driver;
+
+    @FindBy(xpath = "//*[contains(@title, 'Log in to your customer account')]")
```
Selector by title doesn't seem a best one from my pov. Title is more human-readable thing, it can change in time.
Would it be better to have `.header_user_info a` or smth like that instead?

mih-kopylov commented on src/test/java/ui/UserAccountResult.java:
```diff
@@ -0,0 +1,44 @@
+    private String year;
+
+    @Override
+    public boolean equals(Object o) {
```
wow... Why not `@EqualsAndHashCode` from lombok? https://projectlombok.org/features/EqualsAndHashCode

ssheludchenko replied:
I'd like to ask why equals without hashcode?

evgeniyamelone replied:
I think I forgot (

I added annotation.

mih-kopylov commented on src/test/java/ui/UserAccount.java:
```diff
@@ -0,0 +1,43 @@
+public class UserAccount {
+    @NonNull
+    private String firstName;
+    @NonNull
```
if you decide to use `@NonNull` and `@Nulable`, it would be nice to use them in all classes/methods for static analysis.

mih-kopylov commented on src/test/java/ui/ConfigProperties.java:
```diff
@@ -0,0 +1,31 @@
+
+    static {
+        try {
+            fileInputStream = new FileInputStream("src/test/resources/config.properties");
```
Reference to `src` directory doesn't seem correct. When you package the app, there will be no such a directory.
There're two ways of reading properties:
* from classpath
* from external file, and its name should be provided in arguments, like `java --external-config=C:\Projects\config.properties -jar project.jar`

mih-kopylov commented on src/test/java/ui/ConfigProperties.java:
```diff
@@ -0,0 +1,31 @@
+            properties.load(fileInputStream);
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
```
use `try-with-resources` construction to avoid such boilerplate.
```
try(final InputStream is = new FileInputStream("...")){
  //Do whatever you want inside with InputStream
  //The stream will be closed gracefully after the block.
}
```

mih-kopylov commented on src/test/java/ui/ConfigProperties.java:
```diff
@@ -0,0 +1,31 @@
+
+public class ConfigProperties {
+    protected static Properties properties;
+    protected static FileInputStream fileInputStream;
```
It's better to use `InputStream` interface instead of real implementation

mih-kopylov replied:
Is there a reason to have FileInputStream static variable here?

mih-kopylov commented on src/test/java/ui/ConfigProperties.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Properties;
+
+public class ConfigProperties {
+    protected static Properties properties;
```
Is there a reason to have this variable `protected` while there's a `getProperty` method?

mih-kopylov commented on src/test/java/ui/ConfigProperties.java:
```diff
@@ -0,0 +1,31 @@
+            properties = new Properties();
+            properties.load(fileInputStream);
+        } catch (IOException e) {
+            e.printStackTrace();
```
In case of IOException properties won't be read. What's the point of swallowing the exception and letting the application run?
Would it be better to throw this exception wrapped in some runtime one?

mih-kopylov commented on src/test/java/ui/ConfigProperties.java:
```diff
@@ -0,0 +1,31 @@
+    protected static Properties properties;
+    protected static FileInputStream fileInputStream;
+
+    static {
```
nit: I personally don't like static variables and blocks if there's no good reason for them. It's not quite easy to mock them. 

Would it be better to have a regular bean with a normal constructor, etc?

mih-kopylov commented on src/test/resources/config.properties:
```diff
@@ -0,0 +1,3 @@
+mainpage=http://automationpractice.com/index.php
+chromedriver=/Users/evgeniyamaslennikova/Desktop/chromedriver
```
doesn't seem a good idea to put personal configuration to VCS ;) I won't be able to run this test

mih-kopylov commented on src/test/resources/accountData.json:
```diff
@@ -0,0 +1,15 @@
+  "city": "Johannesburg",
+  "state": "10",
+  "postCode": "20450",
+  "mobilePhone": "+8184643447",
```
GDPR! :D 
just kidding

mih-kopylov commented on src/test/java/ui/tests/AuthTest.java:
```diff
@@ -0,0 +1,193 @@
+        accountPage.openPersonalInfo();
+
+        logger.info("Create actual account");
+        UserAccountResult actualAccount =
```
That doesn't seem convenient. It looks like such code should be in some PageObject or smth like that.

ssheludchenko replied:
Agree, if you need to get some information from page, create a method in your page object to avoid `low-level` selenium interactions in test code

mih-kopylov commented on src/test/resources/accountData.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "firstName": "JohnnyUpdated",
```
Just in case: why not using the same `config.properties` file?

ssheludchenko commented on src/test/java/pageobjects/PersonalInformationPage.java:
```diff
@@ -0,0 +1,68 @@
+import ui.UserAccountResult;
+
+@Getter
+@Setter
```
Seems redundant

ssheludchenko commented on src/test/java/ui/UserAccountsPool.java:
```diff
@@ -0,0 +1,62 @@
+
+    public void deserialize(String filePath) {
+        userAccountRegistrationForms = new HashMap<>();
+        ObjectMapper objectMapper = new ObjectMapper();
```
Would be good to have a single object mapper for both `serialize` and `deserialize`, it can have some additional configuration and we don't like to duplicate it

ssheludchenko commented on src/test/java/ui/tests/AuthTest.java:
```diff
@@ -0,0 +1,193 @@
+import java.util.logging.Logger;
+
+public class AuthTest {
+    public static WebDriver driver;
```
Why everything below is `static`?
These could be instance variables, since JUnit will create a new instance of `AuthTest` per `@Test` method

evgeniyamelone replied:
Some of the variables were used  in static methods, I've updated

ssheludchenko commented on src/test/java/ui/tests/AuthTest.java:
```diff
@@ -0,0 +1,193 @@
+    public static AccountPage accountPage;
+    public static PersonalInformationPage personalInformationPage;
+    private static final Logger logger = Logger.getLogger(RegistrationTest.class.getName());
+    Map<String, UserAccount> users = new HashMap<>();
```
Access modifiers?

ssheludchenko commented on src/test/java/ui/UserAccountsPool.java:
```diff
@@ -0,0 +1,62 @@
+    Map<String, UserAccount> userAccountRegistrationForms;
+
+    public void deserialize(String filePath) {
+        userAccountRegistrationForms = new HashMap<>();
```
Why it always re-initialize?

ssheludchenko commented on src/test/java/ui/tests/AuthTest.java:
```diff
@@ -0,0 +1,193 @@
+        UserAccount user = new UserAccount("Johnny", "Flynn", password, email, "14", "3", "1983",
+                "South Africa", "Johannesburg",
+                "10", postCode, mobilePhone, "Home");
+        userAccountsPool.serialize("src/test/resources/accountData.json", user);
```
If it's here for educational purposes this is fine, otherwise it's quite strange that we store something to file and read it immediately

ssheludchenko commented on src/test/java/ui/tests/AuthTest.java:
```diff
@@ -0,0 +1,193 @@
+                .getUser("src/test/resources/accountData.json");
+
+        logger.info("Account registration");
+        accountCreationPage.createAccount(userAccountsPool
```
There's a variable `userPool` (which should be `userData` or `userAccount`) where the content of specified file is already stored

ssheludchenko commented on src/test/java/ui/tests/AuthTest.java:
```diff
@@ -0,0 +1,193 @@
+                        new Select(driver.findElement(By.id("years"))).getFirstSelectedOption().getAttribute("value").trim());
+
+        logger.info("Verify created account");
+        Assert.assertEquals(expectedAccount, actualAccount);
```
If you use `assertEquals` please override `toString` as well, so that assert message will be more readable

evgeniyamelone committed:
Update:
- pom file
- log info

evgeniyamelone committed:
Update:
- remove unneeded getter and setter

evgeniyamelone committed:
Update:
- Implement annotation for equals and hahscode from lombok to compare user account

evgeniyamelone committed:
Update:
- Access modifiers
